### PR TITLE
daemon: make users result more consistent

### PR DESCRIPTION
**Pull request by chipaca**
without this change /v2/users' Result would sometimes be a list of
users, sometimes a single user, and sometimes a bool.

This change makes it always be a list. Should simplify the client
quite a bit.

### CODE REVIEW

File: daemon/api_users.go [hunk 1]
```diff
@@ -342,10 +356,10 @@ func createUser(c *Command, createData postUserCreateData) Response {
 		return InternalError("%s", err)
 	}
 
-	return SyncResponse(&userResponseData{
+	return SyncResponse([]userResponseData{{
```
Comment (mvo5) [inline daemon/api_users.go]: Is this an API break?

Comment (zyga) [reply daemon/api_users.go]: I think, while desired, this is an API break and we should be more creative at to how it is introduced. Perhaps an extra header or some request option. I don't think we can just break this without consequences today.

File: daemon/api_users.go [hunk 2]
```diff
@@ -241,7 +244,27 @@ func postUsers(c *Command, r *http.Request, user *auth.UserState) Response {
 }
 
 func removeUser(c *Command, username string, opts postUserDeleteData) Response {
-	return NotImplemented("not implemented")
```
Comment (zyga) [inline daemon/api_users.go]: Should this whole thing be a change with tasks? What if we refresh snapd in the background and decide to reboot after nuking user from the system but keeping the user in state?

File: daemon/api_users.go (hunk 1 above)
Comment (zyga) [reply daemon/api_users.go]: On deeper look, perhaps it is not? Wasn't createUser just introduced a moment ago?

Comment (chipaca) [reply daemon/api_users.go]: it isn't an API break (unless you count changes between revisions only on master to be such): the `/v2/create-user` endpoint preserves the existing (inconsistent) behaviour. This establishes the new `/v2/users` api to be "always returns a slice of userResponseData" which seems a lot saner to me.

File: daemon/api_users.go (hunk 2 above)
Comment (chipaca) [reply daemon/api_users.go]: that's a bigger change TBH and would apply to create and login and everything in this file as well as remove, except for the GET to /v2/users.
Remember snapd does (or tries to do) a clean shutdown, waiting for clients, so in general it'll wait for sync requests to finish before going away.

File: daemon/api_users.go [hunk 3]
```diff
@@ -276,7 +277,20 @@ func postCreateUser(c *Command, r *http.Request, user *auth.UserState) Response
 	if err := decoder.Decode(&createData); err != nil {
 		return BadRequest("cannot decode create-user data from request body: %v", err)
 	}
-	return createUser(c, createData)
+
+	rsp := createUser(c, createData)
```
Comment (chipaca) [inline daemon/api_users.go]: a way of avoiding the below nasty code would be to give `createUser` a flag to say "do the old thing". Not sure which is better.

File: daemon/api_users.go (hunk 1 above)
Comment (mvo5) [reply daemon/api_users.go]: \o/ thank you!

File: daemon/api_users.go [hunk 4]
```diff
@@ -276,7 +277,20 @@ func postCreateUser(c *Command, r *http.Request, user *auth.UserState) Response
...
+	// backwards compatibility hack
```
Comment (pedronis) [inline daemon/api_users.go]: this comment feels too terse and slightly misplaced, the hack is really the if after, no?

File: daemon/api_users.go [hunk 5]
```diff
@@ -254,7 +277,20 @@ func postCreateUser(c *Command, r *http.Request, user *auth.UserState) Response
...
+	if createData.Email == "" && createData.Known {
```
Comment (pedronis) [inline daemon/api_users.go]: wouldn't: 

```
if  createData.Email != "" {
     // at most one user was created, backward compatibility
     if rsp, ok := ... {
     }
}
```

be more nested but slightly easier to follow?

File: daemon/api_users.go [hunk 6]
```diff
@@ -472,6 +485,10 @@ type postUserCreateData struct {
 	Sudoer       bool   `json:"sudoer"`
 	Known        bool   `json:"known"`
 	ForceManaged bool   `json:"force-managed"`
+
+	// not from JSON: this indicates whether to preserve backwards
+	// compatibility, which results in more clunky return values
+	compat bool
```
Comment (pedronis) [inline daemon/api_users.go]: singleUserResultCompat perhaps, also I would still use

<field name> indicates ...  ; not from JSON

Comment (chipaca) [reply daemon/api_users.go]: done

**Comment by pedronis** [comment]
@chipaca I merged master and solved the conflict (and tweaked the comment a bit more)

**Pull request merged**